File: app/backend/src/routes/calibration.ts
// Forecast resolution + calibration API (Enh 4a). Lets an analyst record which
// scenario actually occurred and scores the recorded forecast (Brier + log score),
// then aggregates calibration across all resolved topics.
import { Elysia, t } from "elysia"
import { getTopic } from "../pipeline/topicManager"
import { dbGetExpertCouncil, dbGetLatestExpertCouncil } from "../db/queries/expert"
import {
  computeScores,
  dbSaveResolution,
  dbGetResolution,
  dbDeleteResolution,
  dbCalibrationSummary,
  type ForecastEntry,
} from "../db/queries/calibration"

export const calibrationRouter = new Elysia()
  // Cross-topic calibration summary (Brier, log score, reliability bins)
  .get("/api/calibration", () => dbCalibrationSummary())

  // Get the resolution for a topic version (or null)
  .get("/api/topics/:id/resolution", ({ params, query }) => {
    if (query.version != null) {
      const version = parseInt(String(query.version))
      if (Number.isNaN(version)) return { resolution: null }
      return { resolution: dbGetResolution(params.id, version) }
    }
    const council = dbGetLatestExpertCouncil(params.id)
    if (!council) return { resolution: null }
    return { resolution: dbGetResolution(params.id, council.version) }
  })

  // Record which scenario actually occurred → compute & store Brier/log score
  .post("/api/topics/:id/resolve", async ({ params, body, set }) => {
    try {
      await getTopic(params.id)
    } catch {
      set.status = 404
      return { message: "Topic not found" }
    }

    // Determine version (explicit, else latest scored)
    let version = body.version
    let council = version != null ? dbGetExpertCouncil(params.id, version) : dbGetLatestExpertCouncil(params.id)
    if (!council) {
      set.status = 400
      return { message: "No scored verdict found for this topic/version to resolve against" }
    }
    version = council.version
    const ranked = council.final_verdict?.scenarios_ranked ?? []
    if (ranked.length === 0) {
      set.status = 400
      return { message: "Verdict has no ranked scenarios" }
    }

    const forecast: ForecastEntry[] = ranked.map(s => ({
      scenario_id: s.scenario_id,
      title: s.title ?? s.scenario_id,
      probability: typeof s.probability === "number" ? s.probability : 0,
    }))

    if (!forecast.some(f => f.scenario_id === body.resolved_scenario_id)) {
      set.status = 400
      return { message: `resolved_scenario_id "${body.resolved_scenario_id}" is not in this verdict's scenarios` }
    }

    const { brier, logScore } = computeScores(forecast, body.resolved_scenario_id)

    dbSaveResolution({
      topic_id: params.id,
      version,
      resolved_scenario_id: body.resolved_scenario_id,
      resolved_at: new Date().toISOString(),
      notes: body.notes ?? null,
      brier_score: brier,
      log_score: logScore,
      forecast,
    })

    return { ok: true, version, brier_score: brier, log_score: logScore }
  }, {
    body: t.Object({
      resolved_scenario_id: t.String(),
      version: t.Optional(t.Number()),
      notes: t.Optional(t.String({ maxLength: 2000 })),
    }),
  })

  // Un-resolve (clear) a topic version's resolution
  .delete("/api/topics/:id/resolution", ({ params, query }) => {
    const council = dbGetLatestExpertCouncil(params.id)
    const version = query.version ? parseInt(String(query.version)) : council?.version
    if (version == null || Number.isNaN(version)) {
      return { ok: false, message: "No version to clear" }
    }
    dbDeleteResolution(params.id, version)
    return { ok: true }
  })

File: server-py/src/dana/api/calibration.py
"""Calibration routes (⇄ TS routes/calibration.ts).

Resolve a topic's forecast against the real-world outcome → Brier + log score, and aggregate
calibration across all resolved topics (reliability curve). The forecast is taken from the
topic version's scored verdict (final_verdict.scenarios_ranked).
"""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from ..db import reads, writers
from ..rigor.calibration import compute_scores, reliability

router = APIRouter()


class ResolveBody(BaseModel):
    resolved_scenario_id: str
    version: int | None = None
    notes: str | None = None


@router.get("/api/calibration")
async def calibration_summary():
    res = await reads.list_resolutions()
    briers = [r["brier_score"] for r in res if r.get("brier_score") is not None]
    logs = [r["log_score"] for r in res if r.get("log_score") is not None]
    return {
        "count": len(res),
        "mean_brier": round(sum(briers) / len(briers), 4) if briers else None,
        "mean_log_score": round(sum(logs) / len(logs), 4) if logs else None,
        "resolutions": res,
        "reliability": reliability(res),
    }


@router.get("/api/topics/{topic_id}/resolution")
async def get_resolution(topic_id: str, version: int = 1):
    return {"resolution": await reads.get_resolution(topic_id, version)}


@router.post("/api/topics/{topic_id}/resolve")
async def resolve(topic_id: str, body: ResolveBody):
    council = await reads.get_expert_council(topic_id, body.version)
    verdict = council.get("final_verdict") if council else None
    if not verdict or not verdict.get("scenarios_ranked"):
        raise HTTPException(status_code=400, detail={"message": "No scored verdict found for this topic/version to resolve against"})
    version = council["version"]
    forecast = [
        {"scenario_id": s.get("scenario_id"), "title": s.get("title"), "probability": s.get("probability", 0)}
        for s in verdict["scenarios_ranked"]
    ]
    if body.resolved_scenario_id not in {f["scenario_id"] for f in forecast}:
        raise HTTPException(status_code=400, detail={
            "message": f'resolved_scenario_id "{body.resolved_scenario_id}" is not in this verdict\'s scenarios'})
    brier, log_score = compute_scores(forecast, body.resolved_scenario_id)
    await _save(topic_id, version, body.resolved_scenario_id, forecast, brier, log_score, body.notes)
    return {"ok": True, "version": version, "brier_score": brier, "log_score": log_score}


@router.delete("/api/topics/{topic_id}/resolution")
async def delete_resolution(topic_id: str, version: int = 1):
    import asyncio
    removed = await asyncio.to_thread(writers.delete_forecast_resolution, topic_id, version)
    return {"ok": removed, "message": None if removed else "No resolution found"}


async def _save(topic_id, version, sid, forecast, brier, log_score, notes):
    import asyncio
    await asyncio.to_thread(writers.save_forecast_resolution, topic_id, version, sid, forecast, brier, log_score, notes)
